Add support for getBlob() in the router

pnarayanan opened the pull request:
Add support for GetBlob operations in the NonBlockingRouter. This resolves issue #296.

Primary reviewers: Siva, Gopal. Others please feel free to take a look.
Expected time: ~2 hours for one pass. Hoping for a wall clock time of ~1 week for this to get merged after iterations.

Some context to help review the change:

- GetBlob has to support getting both simple blobs (legacy blobs and new style blobs that ended up in single chunks) and composite blobs (multi-chunk). Whether a blob is simple or composite is not known until the first chunk is fetched. Therefore, the first chunk is treated as a special chunk within `GetBlobOperation`.
- The operation callback may have to be invoked before the operation completes, as the result of the get operation has to be made available as soon as the first bytes of the data arrives. Additionally, subsequent chunks may be fetched only after earlier chunk data are consumed by the caller (to limit the buffering within the router).
- The result of a GetBlob operation, which is a `ReadableStreamChannel`, is implemented by an inner class (`GetBlobResult`), which is the result passed to the caller.
- Any exceptions encountered after the operation callback is invoked will be notified via the read callback passed in by the caller as part of the `ReadableStreamChannel#readInto()`.
- The `GetManager` will have to continue to poll the `GetBlobOperation` even after the latter invokes the operation callback, because subsequent chunks will have to be fetched in the context of the poll. Additionally, even if all the chunks are fetched, the operation can only complete until the `ReadableStreamChannel#readInto()` is called and all the data is written out. Until that is done, `GetBlobOperation` does not mark itself complete so that it continues to get polled by the `GetManager`.
- The `GetChunk` objects are reused similar to how `PutChunk` objects are reused in `PutOperation`. However, note that chunk buffer allocations do not happen within a `GetChunk`. The `ByteBuffer` is allocated by MessageFormat when deserializing the response.
- Although much of the code in `GetBlobOperation` get exercised in the context of the `RequestResponseHandler` thread, some of it gets exercised in the context of the caller, when it makes `ReadableStreamChannel#readInto()` and in the context of the invoker of the AsyncWritableChannel callback, possibly concurrently.

---

The core changes are in `GetBlobOperation.java`. You will notice while reviewing that not a lot of things are in the abstract `GetOperation` class. This is because of the asymmetry between `GetBlobInfoOperation` and `GetBlobOperation`. For the latter, much of the work related to sending requests and handling response happens within the `GetChunk` objects.

--
*Coverage*
```
GetBlobOperation    100% (6/ 6)    92.7% (38/ 41)    93% (199/ 214)
GetManager          100% (2/ 2)    100% (10/ 10)     90.2% (37/ 41)
GetOperation        100% (1/ 1)    90% (9/ 10)       95.5% (21/ 22)
```
A few more tests will be added.

nsivabalan commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java:
```diff
@@ -60,19 +60,31 @@
+   * {@inheritDoc}
+   */
+  @Override
+  void abort(Exception abortCause) {
```
What is the purpose of the arg here. Also, how can we assume that router close is the only possibility?

pnarayanan replied:
`abortCause` should be used as the exception in the callback. Fixed.

nsivabalan commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java:
```diff
@@ -242,15 +254,45 @@ private void handleBody(InputStream payload)
+   * Process the given {@link ServerErrorCode} and set operation status accordingly.
+   * @param errorCode the {@link ServerErrorCode} to process.
+   */
+  void processServerError(ServerErrorCode errorCode) {
```
private

pnarayanan replied:
done.

nsivabalan commented on ambry-router/src/main/java/com.github.ambry.router/GetOperation.java:
```diff
@@ -141,29 +141,10 @@ boolean isOperationComplete() {
-        break;
-    }
-  }
+  abstract void abort(Exception abortCause);
```
Can you throw some light as to why we need this method. Also, what is the difference between this method and setOperationException

pnarayanan replied:
this is used to abort an operation with an exception. This can be called from within the operation or by the manager when it gets closed abruptly.
`setOperationException` only sets the exception associated with an operation, and is private to the operation. It ensures that the most relevant exception gets set.

nsivabalan commented on ambry-router/src/main/java/com.github.ambry.router/GetOperation.java:
```diff
@@ -44,8 +44,8 @@
   protected final Time time;
-  protected boolean operationCompleted = false;
-  protected Exception operationException;
+  protected volatile boolean operationCompleted = false;
```
Is volatile mandatory? this is updated only during handleResponse() calls if I am not wrong.

pnarayanan replied:
This gets set in an `abort()` call after the operation callback is invoked. You can see that for `GetBlobOperation#abort()`. Will set that for `GetBlobInfoOperation#abort()` as well.

Though it may be okay now to have that as non-volatile in the current code, it seemed important at the time of development that this variable be set at the end of operation completion. Would like to keep it that way.

nsivabalan commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,777 @@
+  // the callback to use to complete the operation.
+  private final OperationCompleteCallback operationCompleteCallback;
+  // whether the operationCompleteCallback has been called already.
+  private final AtomicBoolean operationCallbackCalled = new AtomicBoolean(false);
```
you can consider naming it as "opCompleteCallback" incase you feel "operationCompleteCallback" is too long

pnarayanan replied:
I have renamed the variable to `operationCallbackInvoked`

nsivabalan commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,777 @@
+  private ListIterator<StoreKey> chunkIdIterator;
+  // chunk index to retrieved chunk buffer mapping.
+  private Map<Integer, ByteBuffer> chunkIndexToBuffer;
+  // To find the PutChunk to hand over the response quickly.
```
To find the "GetChunk"

pnarayanan replied:
Thanks, fixed.

nsivabalan commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,777 @@
+      // Get the GetOperation that generated the request.
+      GetRequestInfo getRequestInfo = correlationIdToGetRequestInfo.remove(correlationId);
+      if (getRequestInfo == null) {
+        // Ignore right away. This associated operation has completed.
```
do you think this is valid. We had a consensus that this will not happen incase getBlobInfo

pnarayanan replied:
This is for `GetChunk` and it can absolutely happen. The situation is as follows:

1. `GetChunk` issues two requests with correlation ids 1 and 2.
2.  The first response comes in (say for request 1), and the chunk operation completes, and the chunk's state is cleared, the map itself is cleared.
3. The second response comes in, and it will no longer find correlation id 2 in the map, because the state of the `GetChunk is now different. In this case, the retrieved `GetRequestInfo` will be null.

Note that `GetBlobInfoOperation` does not have this problem as there are no chunks involved.

vgkholla replied:
Won't you have cleared the `correlationIdToGetChunk` entry on chunk completion?

vgkholla replied:
Ah never mind.

nsivabalan commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,777 @@
+      // Although an attempt is made to write to the channel as soon as a chunk is successfully retrieved,
+      // the caller might not have called readInto() and passed in a channel at the time. So an attempt is always
+      // made from within poll.
+      maybeWriteToChannel();
```
How can we write to channel (do IO) during a poll(). Wouldn't this affect our throughput.

pnarayanan replied:
all writes are async.

nsivabalan commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,777 @@
+   */
+  @Override
+  public long getSize() {
+    return -1;
```
Isn't this the size of the blob which we can initialize once we get the first chunk

pnarayanan replied:
We can't, the size is not known unless `BlobProperties` are fetched. If this was necessary, we could have changed the `MessageFormat` for blob content. But all that is unnecessary, this method is not all that useful.

nsivabalan commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,777 @@
+      // read callback.
+      setOperationException(chunk.getChunkException());
+    }
+    if (chunk == firstChunk) {
```
Why do we need to proceed when there is some exception

nsivabalan replied:
lets say we have 4 chunks in flight. and the 4th one resulted in exception. We would write all the remaining chunks to the asyncWritableChannel via maybeWriteToChannel() and then abort the operation since operationException was set due to 4th chunk

pnarayanan replied:
I don't think there is a correctness issue here, but I think we should fail faster like you suggest. Will make the change.

pnarayanan replied:
fixed.

nsivabalan commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,777 @@
+  // the callback that is passed into the asyncWritableChannel write() operation.
+  Callback<Long> chunkAsyncWriteCallback = new Callback<Long>() {
+    @Override
+    public void onCompletion(Long result, Exception exception) {
```
Ideally I would like to rely on this callback and trigger more Get calls or notify that one GetChunk spot is available.

vgkholla replied:
+1. Would that make the code simpler ?

pnarayanan replied:
It seemed logical to do that at first, but then it didn't seem like the right thing to do. Note that this method gets called in the context of another thread. We would then have to worry about making the methods that write out thread safe. It didn't seem worth it, as at the most there would be `MAX_IN_MEM_CHUNKS - 1` or so chunks ready to be written out. And in the current scheme, all of those will get written out in the very next poll.

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java:
```diff
@@ -60,19 +60,31 @@
   }
 
   /**
+   * {@inheritDoc}
```
need not say this I think. It is the default. AFAIK,  {@inheritDoc} is used when you want the parent's doc + you want to add something. If you don't add javadoc at all, it should pull from parent.

pnarayanan replied:
fixed.

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java:
```diff
@@ -60,19 +60,31 @@
+  @Override
+  void abort(Exception abortCause) {
+    operationCompleteCallback.completeOperation(operationFuture, operationCallback, null, abortCause);
+    operationCompleted = true;
```
does the ordering matter? Just curious.

pnarayanan replied:
Perhaps not. But it made sense to do this in case code elsewhere assumed that if `operationCompleted` is true, then the operation was certainly completed. For example, if, for whatever reason, somewhere we chose to do this:

```java
if (operationCompleted) {
  // clear the state of the operation
  operationFuture = null;
  operationCallback = null;
}
```
Though this does not arise today. Was also trying to be consistent with the way we do this for chunks, where we mark the state as ready or complete or free only after all the work is done (which matters for the chunks).

nsivabalan replied:
We do it in reverse order in checkAndMaybeComplete() in the same class

pnarayanan replied:
Yeah, the ordering does not really matter for this class as at the time this variable is set, no other threads check for its status. This was mainly done for `GetBlobOperation` class because there are multiple threads involved there.

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java:
```diff
@@ -182,13 +194,13 @@ void handleResponse(ResponseInfo responseInfo) {
 
   /**
-   * Process the actual GetResponse extracted from a {@link ResponseInfo}
+   * Process the GetResponse extracted from a {@link ResponseInfo}
```
minor: {@link GetResponse}

pnarayanan replied:
done.

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java:
```diff
@@ -182,13 +194,13 @@ void handleResponse(ResponseInfo responseInfo) {
   /**
-   * Process the actual GetResponse extracted from a {@link ResponseInfo}
+   * Process the GetResponse extracted from a {@link ResponseInfo}
    * @param getRequestInfo the associated {@link RequestInfo} for which this response was received.
```
`GetRequestInfo`?

pnarayanan replied:
fixed.

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,777 @@
+ *
+ * This class implements {@link ReadableStreamChannel} and passes itself as the result of this operation.
+ */
+class GetBlobOperation extends GetOperation<ReadableStreamChannel> implements ReadableStreamChannel {
```
Not sure how you feel about this but `GetOperation` and `ReadableStreamChannel` are two very different kinds of functionality. It is like saying an operation is also a channel. 

If you think this is the simplest and easiest way, I'm ok with that but I would have preferred an inner class implementing `ReadableStreamChannel`.

pnarayanan replied:
I don't think `implements` conveys an is-a relationship the way `extends` does. That said, I tried moving the `ReadableStreamChannel` part into an inner class for better encapsulation. However, faced a few issues with that which I can't recollect now. Will attempt doing that once again.

pnarayanan replied:
Done. Introduced an inner class `GetBlobResult` that implements `ReadableStreamChannel`.

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,777 @@
+  // the total number of data chunks retrieved so far (and may or may not have been written out yet).
+  private int numChunksRetrieved;
+  // a list iterator to the chunk ids that need to be fetched for this operation, if this is a composite blob.
+  private ListIterator<StoreKey> chunkIdIterator;
```
wouldn't this belong in `FirstChunk`? (also `BlobIdFactory`)

pnarayanan replied:
had this in the `FirstChunk` initially, but moved it out as it seemed more logical for the operation to iterate over this rather than over the `FirstChunk` or using `firstChunk.nextKeyIndex()` and `firstChunk.nextKey()`

vgkholla replied:
Why wouldn't `firstChunk.nextKeyIndex()` and `firstChunk.nextKey()` be logical?

vgkholla replied:
Do you need the index if it is in the `firstChunk`? 

Will go through the whole patch and maybe we can have a discussion. This isn't too big a deal.

pnarayanan replied:
the other reason was `firstChunk.nextKeyIndex()` will have to do a `listIterator.nextIndex()`, which meant that this method should be called before `firstChunk.nextKey()` which may not be immediately clear. Though that is immediately clear when the `listIterator` is used directly.

pnarayanan replied:
in case the above comment wasn't clear - a call has to be made to get both the next chunk and that chunk's index. We eventually do that by making calls to `listIterator.nextIndex()` and `listIterator.next()` and they have to be done in that order as calling `listIterator.next()` advances the cursor.
This is okay if the call is made using the `listIterator`, as this is its behavior. For `firstChunk`, this may not be immediately clear.

vgkholla replied:
Just making a note here to discuss this with you and make sure I understand the rationale.

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,777 @@
+  // the number of bytes written out to the asyncWritableChannel. This would be the size of the blob eventually.
+  private Long bytesWritten = 0L;
+  // the number of chunks that have been written out to the asyncWritableChannel.
+  private final AtomicInteger numChunksWrittenOut = new AtomicInteger(0);
```
why is this `AtomicInteger` specifically? Will more than one thread update it?

If this is supposed to be `AtomicInteger`, is there a reason for `indexOfNextChunkToWriteOut` not to be?

pnarayanan replied:
`indexOfNextChunkToWriteOut` and most other variables are only accessed int the context of the `RequestResponseHandler` thread, so they don't have to be atomic or volatile.

`numChunksWrittenOut` is accessed both by the `asyncWritableChannel` callback invoker thread and the `RequestResponseHandler` thread. Additionally, we need an ordering guarantee due to the following scenario:

`RequestResponseHandler` thread does this (simplified):
```java
if (numChunksWrittenOut == numChunksTotal) {
  call readIntoCallback with bytesWritten
}
```
and the `asyncWritableChannel`' callback invoker does this:
```java
void onComplete(...) {
  bytesWritten += what_was_written;
  numChunksWrittenOut++;
}
```
which means that `numChunksWrittenOut` should be incremented only *after* `bytesWritten` is updated. So it has to be atomic or volatile. Just made it atomic due to the increment operation (volatile is probably not the right choice).

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,777 @@
+    }
+    if (chunk == firstChunk) {
+      if (operationCallbackInvoked.compareAndSet(false, true)) {
+        // Complete the operation from the caller's perspective, so that the caller can start reading from the
```
"can start reading if there is no exception"? I was a little confused at first.

pnarayanan replied:
added.

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,777 @@
+  void completeRead() {
+    if (readIntoCallbackCalled.compareAndSet(false, true)) {
+      if (operationException.get() != null) {
+        bytesWritten = null;
```
need not do this, it can be the actual bytes written

pnarayanan replied:
Cool. Wasn't sure if they had to be mutually exclusive. Will change.

xiahome commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,777 @@
+   * @return the {@link Future} that will eventually contain the result of the operation.
+   */
+  @Override
+  public Future<Long> readInto(AsyncWritableChannel asyncWritableChannel, Callback<Long> callback) {
```
Who is the caller of this method?

pnarayanan replied:
the caller is the one that made the getBlob() call. The one that waits on the future or gets notified via the callback. The result of the future/callback (this ReadableStreamChannel) is for the caller to use. The caller uses it by reading from it by making this `readInto()` call.

xiahome commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,777 @@
+      if (!successfullyDeserialized) {
+        BlobData blobData = MessageFormatRecord.deserializeBlob(payload);
+        chunkIndexToBuffer.put(chunkIndex,
+            ByteBuffer.wrap(Utils.readBytesFromStream(blobData.getStream(), (int) blobData.getSize())));
```
1. Will the payload get duplicated? In `blobData` the payload is materialized in memory as `ByteBufferInputStream`, and `Utils.readBytesFromStream` create a byte array, which double the memory usage. Would this be necessary?

2. If `BlobData` does not always have its `stream` as a `ByteBuffer`, this method would be blocking?

pnarayanan replied:
You raise a good point. I initially thought that this requires a change in MessageFormat, which is something I noted as part of issue #313. I didn't (and don't) think the change would be big, but it would have to be tested and deployed carefully, and might also affect the coordinator.
However, I think we can do better for now by making minimal changes to MessageFormatRecord. Basically by having it `BlobData` take in a `ByteBufferInputStream` instead of an `InputStream`. I don't think at this point we have to worry about MessageFormat using any other kind of `InputStream`. Will make that change.

Thanks for bringing this up.

pnarayanan replied:
This requires a change to `ByteBufferInputStream`. Have put up #320 for this.

pnarayanan replied:
This is fixed in the latest commit.

xiahome commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,777 @@
+      }
+      if (operationException.get() != null || numChunksWrittenOut.get() == numChunksTotal) {
+        completeRead();
+        operationCompleted = true;
```
Move `operationCompleted = true` into `completeRead()`?

pnarayanan replied:
Will do. Though note that that is not the only scenario when `operationCompleted` will bet set to true.

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,777 @@
+    if (firstChunk.isReady() || firstChunk.isInProgress()) {
+      firstChunk.poll(requestRegistrationCallback);
+    }
+    if (firstChunk.isComplete()) {
```
could this be an `else` case? Or is possible that the poll completes the `firstChunk` ? (I haven't taken a look at that code yet).

pnarayanan replied:
Yes, the `firstChunk` can become complete after the poll and if so we do what needs to be done in the same iteration.

nsivabalan replied:
can you explain how can firstChunk can be complete in this firstChunk.poll(reqRegCallback) if not an exception

pnarayanan replied:
any chunk (including firstChunk) can get completed during a poll, as requests can time out and the operation tracker may consider the chunk operation as complete.

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,777 @@
+  /**
+   * {@inheritDoc}
+   * <br>
+   * @param asyncWritableChannel the {@link AsyncWritableChannel} to read the data into.
```
is this required? The docs at the interface are not enough?

pnarayanan replied:
Yes, in fact this is exactly the ones in the interface I think. We only need a custom one for `getSize()` which will always return -1.

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,777 @@
+
+  /**
+   * {@inheritDoc}
+   * @return true, unless the channel is closed.
```
same here.

pnarayanan replied:
fixed.

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,777 @@
+
+  /**
+   * {@inheritDoc}
+   * @throws IOException
```
and here.

pnarayanan replied:
fixed.

nsivabalan commented on ambry-router/src/main/java/com.github.ambry.router/GetOperation.java:
```diff
@@ -172,9 +153,9 @@ void processServerError(ServerErrorCode errorCode) {
    */
   void setOperationException(RouterException exception) {
-    if (operationException == null || exception.getErrorCode() == RouterErrorCode.BlobDeleted
+    if (operationException.get() == null || exception.getErrorCode() == RouterErrorCode.BlobDeleted
```
May I know why we over-ride if operationException is not set(i.e, if its set to null). Also, can you also explain what happens when all replicas throws InternalServerError for some reason, what will be the reconciled error code we get ?

pnarayanan replied:
Why wouldn't we? That is the unambiguous case - when the first exception comes in we simply set it no matter what.

Note that setting this exception does not necessarily mean the operation has completed. This method is called every time an exception is received for a request, we may continue to do more requests and this method can get called over and over. Eventually the most relevant exception gets set. If `Blob_Deleted` or `Blob_Expired` is received, the operation completes immediately. If a successful response comes in, `operationException` is cleared before the operation is completed.

nsivabalan replied:
sorry, don't know why did I even ask this question. my bad

nsivabalan commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,777 @@
+        // periodically. If any exception is encountered while processing subsequent chunks, those will be notified
+        // during the channel read.
+        ReadableStreamChannel result = getOperationException() == null ? this : null;
+        operationCompleteCallback
```
So this effectively means that the getOperationsCount() in our NBR will never consider the getBlobOpertions() currently being served as this line will intimate NBR as operation completed. Am I right. Did you give it a thought abt it.

pnarayanan replied:
You are right. I don't think there is much we can do about it though. Internally, for metrics, the operation managers could report the count and that will account for all operations. This variable is in the router, and as far as the router is concerned, the operation will be complete. We could see when we add metrics whether this variable has any value. Currently this is just the number of operations that are waiting for a result from the router.

nsivabalan commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,777 @@
+    if (operationCompleted) {
+      return;
+    }
+    if (firstChunk.isReady() || firstChunk.isInProgress()) {
```
this is also equivalent to checking  ! firstChunk.isComplete() if I am not wrong

nsivabalan replied:
This makes me think if we need to state Ready and InProgress or can we have just one state. For all getChunks() you do something if its in either of these two states (isReady(), IsInProgress()) and we don't we anything in isolation.

nsivabalan replied:
oops, my bad. I didn't see the line 285 completely. ignore the 2nd comment.

pnarayanan replied:
Yes for the first comment only because `firstChunk` is born ready (never goes into `Free` state).

nsivabalan commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,777 @@
+      maybeWriteToChannel();
+      // If this is a composite blob, poll for requests for subsequent chunks.
+      if (dataChunks != null) {
+        for (GetChunk dataChunk : dataChunks) {
```
should we fail fast to check for operationException at places like here. Since an exception can be thrown via chunkAsyncWriteCallback, thought no point in iterating over the dataChunks.

nsivabalan replied:
Also, can you help me understand what happens in this case or is my understanding right. Lets say we poll for the data chunks here and have collected all the required requests that needs to be sent out, but just before returning an exception was thrown via chunkAsyncWriteCallback and we complete the operation. But then, we have registered all the requests with the requestRegistrationCallback which means that the NBR will send out those requests which is of no use. Only during next poll(), we won't revisit this operation again, coz GetManager.poll() would have removed this GetOperation from its list since the operation is complete.

nsivabalan replied:
Haven't gone through Put implementation. Might be worth looking into other operations as well.

pnarayanan replied:
Those exceptions could come in just after we returned from the poll as well and there is nothing that can be done in that case. I think this is a small price to pay. You could think of registering a request to be as good as having sent it.

For your first comment, yes, I think we should check for the operation status after each chunk is polled. We do that for `PutOperation`. Will make a similar change.

nsivabalan commented:
done with one pass for production code. Haven't given much attention to documentation yet. Will do it next pass along with tests.

vgkholla commented:
production code done. Will review tests tomorrow.

pnarayanan commented:
Addressed all comments so far. Please take a look.

pnarayanan commented:
@vgkholla I think your idea of reviewing the tests separately is good and will help get this merged sooner. Newly added tests are isolated to `GetBlobOperationTest.java` and `GetManagerTest.java` and have created #329 to improve those. When the patch for that gets reviewed, you could go over the whole test. Alternatively, the tests here could be reviewed after merging and any comments can be incorporated as part of the fix for #329.

fyi, the coverage is > 90% currently.

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,807 @@
+ * buffered up to the maximum that can be buffered. When fetched chunks are consumed by the caller, subsequent chunks
+ * become eligible to be fetched.
+ *
+ * This class implements {@link ReadableStreamChannel} and passes itself as the result of this operation.
```
Minor: this is no longer true, right?

vgkholla replied:
The PR description might also require a rewrite of this line
> The result of a GetBlob operation, which is a ReadableStreamChannel, is implemented directly by the GetBlobOperation, which makes sharing states easy.

pnarayanan replied:
fixed.

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,807 @@
+      throws RouterException {
+    super(routerConfig, routerMetrics, clusterMap, responseHandler, blobIdStr, futureResult, callback, time);
+    this.operationCompleteCallback = operationCompleteCallback;
+    this.blobIdFactory = new BlobIdFactory(clusterMap);
```
minor: curious if this can be passed from the `GetManager` (or whoever is creating this object) instead of creating a new object for every `GetBlobOperation`?

pnarayanan replied:
That is a good idea. Done.

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,807 @@
+   * appropriately.
+   * @param chunk the chunk that has completed.
+   */
+  private void onChunkOperationComplete(GetChunk chunk) {
```
minor: Why isn't this a part of the code in `GetChunk`? (con: there might be some repeated code if it is moved there).

nsivabalan replied:
+1. actually might look simplified. if(chunk == firstChunk) block will be present only in the FirstChunk implementation and the if loop is not required.

pnarayanan replied:
It does things like setting operation exception and such which should be done by the operation. It also creates the result object and accesses other things which are not chunk specific. That way it seemed better done within the operation. This is also consistent with puts.

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,807 @@
+    // the index of the next chunk that is to be written out to the asyncWritableChannel.
+    private int indexOfNextChunkToWriteOut = 0;
+    // whether this object has called the readIntoCallback yet.
+    private AtomicBoolean readIntoCallbackCalled = new AtomicBoolean(false);
```
`final`?

pnarayanan replied:
done.

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,807 @@
+    // whether this object has called the readIntoCallback yet.
+    private AtomicBoolean readIntoCallbackCalled = new AtomicBoolean(false);
+    // the callback that is passed into the asyncWritableChannel write() operation.
+    Callback<Long> chunkAsyncWriteCallback = new Callback<Long>() {
```
`private final`?

pnarayanan replied:
done.

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,807 @@
+        if (exception != null) {
+          operationException.set(exception);
+        }
+        numChunksWrittenOut.incrementAndGet();
```
I might have asked this before but still not sure why this needs to be an `AtomicInteger`? Should be enough if it is `volatile`, just like `bytesWritten`?

pnarayanan replied:
`bytesWritten` is not accessed concurrently. It is accessed by two thread, but not concurrently.
`numChunksWrittenOut` is accessed concurrently. `volatile` guarantees ordering, but not atomicity of increment operations. `AtomicInteger` guarantees both.

vgkholla replied:
Not sure why atomicity of increment operations is important here. There is only ever one thread doing that?

pnarayanan replied:
Did some reading and it seems volatile would be good enough here.
In general, volatile is not sufficient if modifying the variable requires reading the same variable. However, if only one thread is ever modifying the variable, then volatile seemingly suffices as per this [article](http://www.ibm.com/developerworks/library/j-jtp06197/).

vgkholla replied:
Thanks for following this up ! That article is useful to understand use cases of `volatile`.

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,807 @@
+
+    @Override
+    public Future<Long> readInto(AsyncWritableChannel asyncWritableChannel, Callback<Long> callback) {
+      if (readCalled) {
```
minor: although i don't envision a use case, the interface requires this to be thread safe i.e. if `readInto()` is called concurrently, only one should win.

pnarayanan replied:
I don't think the interface should be thread safe. In general, I feel that one must have a good reason to make interfaces thread safe as that can burden the implementations unnecessarily. If a caller wants concurrency, then perhaps another "concurrent" interface can extend from the original interface. It doesn't seem worthwhile to have all implementations be thread-safe even if callers don't need them to.

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,807 @@
+      if (firstChunk.isComplete() && readCalled) {
+        while (operationException.get() == null && chunkIndexToBuffer.containsKey(indexOfNextChunkToWriteOut)) {
+          ByteBuffer chunkBuf = chunkIndexToBuffer.remove(indexOfNextChunkToWriteOut);
+          asyncWritableChannel.write(chunkBuf, chunkAsyncWriteCallback);
```
Might be an orthogonal question, but if this (or any other statement really) were to throw a `RuntimeException`, who would catch and what is the behavior? And is that the right behavior?

pnarayanan replied:
If that happens, it will bring down the router today. Although `RunTimeException` should not ideally happen unless there is a programming error, we can discuss on what the right behavior should be. In the worst case, we could have a catch all within the operation managers that simply removes the bad operation from its list.

vgkholla replied:
Certainly should not bring down the `Router` IMO. We should have the catch-all since anyone may provide an implementation of `AsyncWritableChannel`.

pnarayanan replied:
Yes. Added this to my backlog. Basically, we need to deal with exceptions better for both the ChunkFiller thread and the RequestResponseHandler thread so that one operation's bad state only affects that operation and not the entire router.

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,807 @@
+     * Do what needs to be done after the GetBlobOperation is done with the current chunk that has completed. When
+     * this is called, the state of the chunk should be {@link ChunkState#Complete}
+     */
+    void postCompletionCleanup() {
```
minor: do we need this function? Just the `reset()` will do IMO.

pnarayanan replied:
this gets overridden in the `FirstChunk`

vgkholla replied:
why not override `reset()`?

pnarayanan replied:
`FirstChunk` is also a `GetChunk` and `reset()` is part of `GetChunk` constructor.

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,807 @@
+    // the actual body of the response is deserialized only once.
+    protected boolean successfullyDeserialized;
+    // map of correlation id to the request metadata for every request issued for this operation.
+    protected final Map<Integer, GetRequestInfo> correlationIdToGetRequestInfo = new TreeMap<>();
```
minor: Would `LinkedHashMap` be a better candidate here? Space v/s time tradeoff.

pnarayanan replied:
that is a good choice too. Doesn't make much of a difference here as this data structure will never grow much (bounded by the number of replicas)

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,807 @@
+   */
+  enum ChunkState {
+    /**
+     * The GetChunk is free and can be assigned to hold a chunk of an object.
```
object seems vague. Would "composite blob" be better?

pnarayanan replied:
will use "overall blob" like I've used elsewhere (it does not necessarily have to be a composite blob).

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/GetManager.java:
```diff
@@ -161,8 +153,9 @@ void handleResponse(ResponseInfo responseInfo) {
-        onComplete(getOperation);
+      if (getOperation.isOperationComplete()) {
+        getOperations.remove(getOperation);
+        routerMetrics.operationDequeuingRate.mark();
```
is it possible you will double count if `close()` has removed the operation already?

pnarayanan replied:
yeah, not that it is a big deal if the metric gets counted during close, but since this can be easily handled, will make the change. Introduced a method to remove from `getOperations`. Anytime we remove, we will need to update this (and the get rate metric which will come in later).

vgkholla commented on ambry-router/src/test/java/com.github.ambry.router/GetBlobInfoOperationTest.java:
```diff
@@ -149,7 +154,7 @@ public void onCompletion(BlobInfo result, Exception exception) {
     GetBlobInfoOperation op =
         new GetBlobInfoOperation(routerConfig, routerMetrics, mockClusterMap, responseHandler, blobIdStr,
-            operationFuture, operationCallback, time);
+            operationFuture, operationCallback, operationCompleteCallback, time);
```
why is ` operationsCount.incrementAndGet()` not required in these two cases?

vgkholla replied:
In this case only actually. For the other one, the object isn't even constructed so that's ok.

pnarayanan replied:
because this operation is not taken to completion for the decrement to happen (unlike the rest of the operations).

pnarayanan replied:
added a comment here - I realize it isn't very intuitive.

vgkholla commented on ambry-router/src/test/java/com.github.ambry.router/GetManagerTest.java:
```diff
@@ -133,6 +136,34 @@ public void testFailureOnAllPollThatSends()
+    do {
+      ByteBuffer buf = getChannel.getNextChunk();
+      while (buf.hasRemaining()) {
+        Assert.assertEquals("Get and Put blob content should match", putContent[readBytes++], buf.get());
```
this will throw an `IndexOutOfBoundsException` if `putContent` runs out but `buf` has not. While you could argue that there is no problem since the test will fail and we will eventually debug, I would find it more informative if the test could check that the index has gone out of range and fail with a helpful message indicating the root cause i.e unexpected "extra" data returned.

pnarayanan replied:
Done.

vgkholla commented on ambry-router/src/test/java/com.github.ambry.router/GetManagerTest.java:
```diff
@@ -133,6 +136,34 @@ public void testFailureOnAllPollThatSends()
+      }
+      getChannel.resolveOldestChunk(null);
+    } while (readBytes < putContent.length);
   }
```
I think that there are two final checks required to ensure that there is no "extra" data (in the same order).

1. The future returned from the `readInto()` should be "done".
2. There should be no more chunks to read (`assertNull("There was unexpected extra data", getChannel.getNextChunk(0))`).

nsivabalan replied:
+1. I made the same comment earlier in other test class

pnarayanan replied:
Fixed, in a slightly different way.

vgkholla commented:
Made a pass through everything except `GetBlobOperationTest`.

For the code I have reviewed, minor comments. LGTM otherwise.

Will review `GetBlobOperationTest` while you wait on others to review and give a ship-it but if others are done, let's merge and I will review and provide comments on the test that you can address while adding more tests.

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,807 @@
+      public void onCompletion(Long result, Exception exception) {
+        bytesWritten += result;
+        if (exception != null) {
+          operationException.set(exception);
```
leaving a note here to discuss stomping of exceptions.

nsivabalan commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,807 @@
+     * @param errorCode the {@link ServerErrorCode} to process.
+     */
+    void processServerError(ServerErrorCode errorCode) {
+      setChunkException(new RouterException("Server returned: " + errorCode, RouterErrorCode.UnexpectedInternalError));
```
Why is this so? Any error is considered to be UnexpectedInternalError ?

pnarayanan replied:
Yes. Also see the documentation for this method in `FirstChunk`, which I have now also added to the docs here.

If a data chunk is being retrieved, that means the first chunk was retrieved successfully, and the blob is "good", so any error is unexpected. We can't return a blob deleted error for a chunk at this point, for example, as the operation callback was already returned with "blob found".

nsivabalan commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -0,0 +1,807 @@
+
+    @Override
+    GetOptions getGetOptions() {
+      return GetOptions.None;
```
Shouldn't this is passed in arg to GetBlobOperation. In future, admin will be used to retrieve expired blobs, deleted blobs and so on which will use the router to fetch blobs

pnarayanan replied:
We don't have any support for that at this time. The router does not expose any "per operation" flags. If/when we need those functionalities, we will have to introduce a new version of the router api.

nsivabalan commented:
Few minor comments in production code. I will make sure I will review tests by this evening.

nsivabalan commented on ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java:
```diff
@@ -0,0 +1,596 @@
+        }
+        asyncWritableChannel.resolveOldestChunk(null);
+      } while (readBytes < blobSize);
+      written = readFuture.get();
```
What incase we have more data in the channel ? Shouldn't we assert this.

vgkholla replied:
Yeah same two checks as described in the other comment of mine would be good here I think.

pnarayanan replied:
Fixed.

nsivabalan commented on ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java:
```diff
@@ -0,0 +1,596 @@
+  @Test
+  public void testCompositeBlobNotChunkSizeMultipleGetSuccess()
+      throws Exception {
+
```
why extra line

pnarayanan replied:
fixed.

nsivabalan commented on ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java:
```diff
@@ -0,0 +1,596 @@
+            null, operationCompleteCallback, time);
+    requestRegistrationCallback.requestListToFill = new ArrayList<>();
+    op.poll(requestRegistrationCallback);
+    while (!op.isOperationComplete()) {
```
Feels bit hacky. Can't we set the error code that the MockServer can throw for the requests

vgkholla replied:
@nsivabalan On the contrary, this is exactly the way this should be written don't you think? Why would a `MockServer` say "timed out at the router"? That feels hacky.

nsivabalan commented on ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java:
```diff
@@ -0,0 +1,596 @@
+    routerConfig = new RouterConfig(new VerifiableProperties(props));
+    testVariousErrors(dcWherePutHappened);
+
+    props = getNonBlockingRouterProperties();
```
why do we need this line.

pnarayanan replied:
why do we not?

nsivabalan commented on ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java:
```diff
@@ -0,0 +1,596 @@
+    routerConfig = new RouterConfig(new VerifiableProperties(props));
+    testVariousErrors(dcWherePutHappened);
+
+    props = getNonBlockingRouterProperties();
```
same here

vgkholla commented on ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java:
```diff
@@ -0,0 +1,596 @@
+  private int successTarget = 1;
+  private int maxChunkSize;
+  private int blobSize = 1000;
+  private RouterConfig routerConfig;
```
minor: a lot of these could be `final` and instantiated in the constructor?

pnarayanan replied:
the idea is to construct the router after setting config variables custom to the tests. So the router and the properties have to be constructed outside of the router. Since `doPut()` is the first thing each test does, we are constructing all these there.

vgkholla commented on ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java:
```diff
@@ -0,0 +1,596 @@
+  private byte[] userMetadata;
+  private byte[] putContent;
+
+  private final GetTestRequestRegistrationCallbackImpl requestRegistrationCallback =
```
minor: implementation need not be declared as separate class I think. It could be defined inline.

vgkholla replied:
or is it because you need to clear `requestList`? If that is the case, could you make that variable non-private. I know it doesn't matter accessibility-wise but while reading it makes it easier to understand it is used externally.

pnarayanan replied:
yes, and done.

vgkholla commented on ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java:
```diff
@@ -0,0 +1,596 @@
+    mockClusterMap = new MockClusterMap();
+    routerMetrics = new NonBlockingRouterMetrics(mockClusterMap.getMetricRegistry());
+    mockServerLayout = new MockServerLayout(mockClusterMap);
+    replicasCount = mockClusterMap.getWritablePartitionIds().get(0).getReplicaIds().size();
```
minor: this seems a bit awkward to me (even though the cluster map has only one partition). It seems more natural to me get the count from the blob id instead.

vgkholla replied:
and clearer. Making this a class variable was a little confusing for me.

pnarayanan replied:
The assumption is that all partitions have the same number of replicas. That is a fair assumption I think, as things like `successTarget` and `requestParallelism` are not partition specific.
That way, I think it belongs in the class and does not have to be computed on the fly.

vgkholla commented on ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java:
```diff
@@ -0,0 +1,596 @@
+   */
+  private void doPut()
+      throws Exception {
+    requestParallelism = 3;
```
not sure what you are trying to do here.

pnarayanan replied:
will refactor this a bit.

vgkholla commented on ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java:
```diff
@@ -0,0 +1,596 @@
+   * Recreate the configs for the router to be used by gets.
+   */
+  private void resetConfigsForGet() {
+    requestParallelism = 2;
```
what are you resetting here? Don't put and get have different config params anyway? Looks like I'm missing something.

vgkholla replied:
The chunk size is being rest, that much is clear, but the parallelism and success target?

pnarayanan replied:
yeah, this is not right, sorry. will refactor.

vgkholla commented on ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java:
```diff
@@ -0,0 +1,596 @@
+    putContent = new byte[blobSize];
+    random.nextBytes(putContent);
+    ReadableStreamChannel putChannel = new ByteBufferReadableStreamChannel(ByteBuffer.wrap(putContent));
+    blobIdStr = router.putBlob(blobProperties, userMetadata, putChannel).get();
```
this is a personal preference, but I really dislike too many class variables in tests. It makes reading (and reviewing) the tests so complicated. Why can't `blobIdStr` be returned from the `doPut()`?

pnarayanan replied:
There is a tradeoff. The alternative is for all these methods to share state via return values (like what you suggest) or pass them as arguments. Which is not any better either.

The only value is if variables can be local to a method - in which case they should not be made members. Otherwise, if different methods are sharing the state, might as well make them members.

For this particular example, if `doPut()` could return the `blobIdStr`, but then certain tests will have to pass it down to helper methods. If the variable gets passed across 2-3 methods of the class, why not simply make it a member variable?

vgkholla commented on ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java:
```diff
@@ -0,0 +1,596 @@
+  private void resetConfigsForGet() {
+    requestParallelism = 2;
+    successTarget = 1;
+    VerifiableProperties vprops = new VerifiableProperties(getNonBlockingRouterProperties());
```
also same thing with using class variables in this function -> `getNonBlockingRouterProperties()`. It is very unintuitive (at least for me) to set certain variables that will influence this function when it might be more readable to just pass them as arguments and isolate the function.

pnarayanan replied:
some of what you say makes sense. will refactor this.

vgkholla commented on ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java:
```diff
@@ -0,0 +1,596 @@
+    maxChunkSize = random.nextInt(1024 * 1024) + 1;
+    blobSize = maxChunkSize * 5;
+    doPut();
+    String blobIdStr = (new BlobId(mockClusterMap.getWritablePartitionIds().get(0))).getID();
```
minor: there is already a `blobIdStr` that was set by the `doPut()`. Why the new one?

vgkholla replied:
Also not sure why the put was required. Do you get it somewhere?

vgkholla commented on ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java:
```diff
@@ -0,0 +1,596 @@
+    try {
+      new GetBlobOperation(routerConfig, routerMetrics, mockClusterMap, responseHandler, "invalid_id", operationFuture,
+          operationCallback, operationCompleteCallback, time);
+      Assert.fail("Instantiation of GetReadableStreamChannel operation with an invalid blob id must fail");
```
`GetReadableStreamChannel`?

pnarayanan replied:
fixed.

vgkholla commented on ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java:
```diff
@@ -0,0 +1,596 @@
+   * Instantiate a router, perform a put, close the router. The blob that was put will be saved in the MockServer,
+   * and can be queried by the getBlob operations in the test.
+   */
+  private void doPut()
```
would it be possible to order it such that all tests (public functions) are on the top and all helpers (private functions), at the bottom?

pnarayanan replied:
This ordering made sense to me. `doPut()` is something every test does at the beginning. It is a bit like `Before`, except that there are chunk and blob sizes to set before calling it. The other private methods usually are arranged closer to the public methods that call them. I generally like that as that makes  easy to scan through code.

Will see if there is anything that I can improve on.

vgkholla commented on ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java:
```diff
@@ -0,0 +1,596 @@
+    for (int i = 0; i < 10; i++) {
+      // blobSize in the range [1, maxChunkSize]
+      blobSize = random.nextInt(maxChunkSize) + 1;
+      doPut();
```
once again, personal pref - but i would find the tests *much* more readable if `doPut()` just took the arguments it needed instead of us setting the arguments via class variables. 

One of the reasons I say that is because if I want to reuse the function and want to know what its knobs are, I *have* to read the whole function. If it took arguments, I would simply read the description of the arguments and know what I need.

Any new arguments will also need to be added as class variables and that might make it more confusing.

pnarayanan replied:
will see how to refactor. If `doPut()` were the constructor, then what you say does not apply - it couldn't take any arguments under junit. In fact, it was initially the constructor, but later I wanted certain tests to be able to set certain values before the things in the constructor executed, and that is why this ended up being a different method.

vgkholla commented on ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java:
```diff
@@ -0,0 +1,596 @@
+      }
+    }
+
+    Assert.assertEquals("Must have attempted sending requests to all replicas", replicasCount,
```
this looks like its been repeated thrice. Perhaps calls for a utility function? Especially since the same test can be made for internal server errors, IO exceptions etc.

pnarayanan replied:
Just couple of asserts, perhaps not worth creating a different method. I find this more readable. Sometimes having too many utility methods for such things make things unreadable I feel - hard to get the context from the method itself without looking at the usage.

vgkholla commented on ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java:
```diff
@@ -0,0 +1,596 @@
+          operationException.set(exception);
+          readCompleteLatch.countDown();
+        } else {
+          Utils.newThread(new Runnable() {
```
not convinced you need a different thread for this. Could you explain?

Just seems to be complicating the code and might complicate debugging - too many moving parts.

vgkholla replied:
Not sure if I'm entirely correct but I suspect that if is possible to get rid of this thread, then you would avoid waiting on latches because everything would be done synchronously.

vgkholla replied:
Ok I think I see why the thread is needed - because the data needs to be read and that is blocking.

vgkholla commented on ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java:
```diff
@@ -0,0 +1,596 @@
+
+    readCompleteLatch.await();
+    Assert.assertTrue("Operation should be complete at this time", op.isOperationComplete());
+    Assert.assertNull("Operation exception should be null, was:" + operationException.get(), operationException.get());
```
why not this?
```
if(operationException.get() != null) {
 throw operationException.get();
}
```

In the current way, you lose the stacktrace (same thing with `readCompleteException`).

pnarayanan replied:
makes sense.

vgkholla commented:
Finished a full pass.

pnarayanan commented:
Thanks for the reviewing so far guys. Have addressed all your comments for the main code, and most for the tests. Will be refactoring the tests a bit based on your comments. Will update the patch tomorrow.

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -115,11 +113,12 @@
       ResponseHandler responseHandler, String blobIdStr, FutureResult<ReadableStreamChannel> futureResult,
-      Callback<ReadableStreamChannel> callback, OperationCompleteCallback operationCompleteCallback, Time time)
+      Callback<ReadableStreamChannel> callback, OperationCompleteCallback operationCompleteCallback,
+      BlobIdFactory blobIdFactory, Time time)
```
need to add this to the doc.

pnarayanan replied:
fixed.

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/GetManager.java:
```diff
@@ -127,6 +130,21 @@ void submitGetBlobOperation(String blobId, FutureResult<ReadableStreamChannel> f
+   * @param op the {@link GetOperation} to remove.
+   * @return true if the operation was removed in this call.
+   */
+  boolean remove(GetOperation op) {
```
why is this not `private`?

pnarayanan replied:
should be private. fixed.

vgkholla commented on ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java:
```diff
@@ -52,43 +53,47 @@
   private static final int CHECKOUT_TIMEOUT_MS = 1000;
-  private static final int BLOB_USER_METADATA_SIZE = 10;
-
+  private final int REPLICAS_COUNT;
```
caps are only when a variable is both `static` and `final`. Same for the variable below.

pnarayanan replied:
fixed.

vgkholla commented on ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java:
```diff
@@ -0,0 +1,592 @@
+          e.getErrorCode());
+    }
+
+    doPut();
```
do you even need to do this? Can just generate a `blobIdStr` because it looks like that is all you need?

pnarayanan replied:
In this particular case, yes. Seemed simpler to just call this method at the time. Will instead create the blobId directly if that is more intuitive.

vgkholla commented:
Minor comments. LGTM. Is everything built and formatted?
(I have 4 comments - for some reason, two of them say "outdated diff" but please take a look).

Two action items that I remember are about handling runtime exceptions and about adding more tests (not just to increase line coverage but to cover all cases). I think you are tracking both but just want to reiterate.

pnarayanan commented:
Yes. Issues are already in place to track the two. #333 and #329 respectively.

pnarayanan commented:
All comments addressed. Built on Linux and OS X. Formatted.

pnarayanan commented:
If and when this looks good, please squash and merge, keeping and only keeping the subtext from the last commit that starts with "A GetBlobOperation makes use of ...".
Please don't include any line starting with a `*` - not even the first - so that the commit is clean and readable.

nsivabalan commented on ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java:
```diff
@@ -0,0 +1,593 @@
+    networkClient = networkClientFactory.getNetworkClient();
+  }
+
+  private void doPut()
```
java doc

pnarayanan replied:
added.

nsivabalan commented on ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java:
```diff
@@ -0,0 +1,593 @@
+        int bufLength = buf.remaining();
+        Assert.assertTrue("total content read should not be greater than length of put content",
+            readBytes + bufLength <= putContent.length);
+        while (buf.hasRemaining()) {
```
bufLength

nsivabalan commented on ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java:
```diff
@@ -0,0 +1,593 @@
+  public void testErrorPrecedenceWithBlobDeletedAndExpiredCase()
+      throws Exception {
+    doPut();
+    ServerErrorCode[] serverErrorCodesToTest = {ServerErrorCode.Blob_Deleted, ServerErrorCode.Blob_Expired};
```
Defining two arrays is not very apparent of their mapping. May be a map of server error code to router error code will be very intuitive.

pnarayanan replied:
done.

nsivabalan commented on ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java:
```diff
@@ -0,0 +1,593 @@
+    responseList.addAll(networkClient.sendAndPoll(requestList));
+    requestList.clear();
+    while (responseList.size() < sendCount) {
+      responseList.addAll(networkClient.sendAndPoll(requestList));
```
not related to this patch. but shouldn't networkClient.sendAndPoll() support null. In most of our tests, we have to send an empty list.

pnarayanan replied:
Yeah it could. For the main code, that use case does not arise, as the `OperationController` gets a valid list after the poll and that is the list passed to this method. Will defer doing this for now.

nsivabalan commented on ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java:
```diff
@@ -0,0 +1,593 @@
+    // are final), except for one of the servers in the datacenter where the put happened (we do this as puts only go
+    // to the local dc, whereas gets go cross colo).
+    serverErrors.remove(ServerErrorCode.Blob_Deleted);
+    serverErrors.remove(ServerErrorCode.Blob_Expired);
```
guess we also need to remove No_Error

pnarayanan replied:
yes, thanks for catching that.

pnarayanan commented:
Addressed all comments. Built and formatted.
Once again, when squashing and merging, please do the following:
Keep only the subtext from the commit that starts with "A GetBlobOperation makes use of ...".
Please don't include any line starting with a * - not even the first - so that the commit is clean and readable.
